Alerting: Move legacy alert migration from sqlstore migration to service

JacobsonMT opened the pull request:
This PR moves the legacy alert migration out of the sqlstore db migrations and into ngalert.

There are multiple reasons behind this change, I'll do my best to explain them here:

### Legacy alert migration doesn't fit the definition of a sqlstore migration

Initially, this was a natural place for the migration code to live. It does, in fact, migrate database schema and contents. However, it breaks two key requirements of sqlstore migrations:

1. It is not guaranteed to run at a specific point in time relative to other migrations.
2. It changes over time.

These two differences are caused by the fact that the legacy migration is currently optional. Users may decide when to migrate, or even rollback the migration and try again later. This optional status also means we were able to improve the migration over time, thus causing the migration to change.

This has contributed to numerous bugs in the past, either directly and indirectly. To name a few recent ones:

- https://github.com/grafana/grafana/issues/71529 
- https://github.com/grafana/grafana/issues/67448 
- https://github.com/grafana/grafana/issues/60196 
- https://github.com/grafana/grafana/issues/65839 

### Legacy provisioned alerts cannot currently be migrated in stateless grafana deployments

Since sqlstore migrations run before any provisioning, this means that the migration code does not see that legacy provisioned alerts exist.

- https://github.com/grafana/grafana/issues/43546
- https://github.com/grafana/grafana/issues/52388

By moving this into `ngalert.Run()` we allow a followup PR to run dashboard provisioning before we start the legacy migration.

### The migration code is saving vendored models to the database

This makes some sense if the migration were a true sqlstore migration where followup migrations would update the vendored models to more current ngalert versions. This is not currently what is happening. Instead, we periodically update the vendored models in the main migration in an attempt to keep the vendor in sync with ngalert models.

This is prone to drift and is not obvious in the context of sqlstore migrations.

With the migration code part of the ngalert package, a followup PR can ensure it uses the real ngalert models and drift should be caught alongside any other ngalert code changes we do. 

### This unlocks features that were not easily done before

If the migration code is in ngalert as a service, then we are free to use it in new/existing endpoints. This opens up interesting feature possibilities. For example (not saying we will do all of these and some are mutually exclusive):

- No longer migrate legacy alerts on startup, instead make it an endpoint with UI.
  - This button could show you the yaml of all the UA resources that would be created after the migration. The yaml could then be imported into another instances for testing in dev, for example.
  - Or maybe, the UI will let you live toggle between legacy and UA as we migrate/revert over and again as you check out the differences.
- Create an endpoint that can take legacy dashboard yaml containing alerts and migrate them ad-hoc directly into UA (or maybe convert to UA yaml)
- Migrate provisioned legacy alerts to provisioned UA alerts with a special provenance (provenance `legacy_migration` or something). As opposed to most alerts, these `legacy_migration` provenance alerts can be overwritten on a subsequent migration run. This could allow us to let users keep their as-code provisioned alerts in legacy format for a while if desired.


Edit:

For the purposes of keeping the change self-contained in a single commit, this PR has been combined with a [2nd PR](https://github.com/grafana/grafana/pull/74503) to replace the vendored models in the migration with their equivalent ngalert models. It also replaces the raw SQL selects and inserts with service calls:

- Dashboard Alerts: 9f5b6b2ed2efb9f6bfeb41de7c1172938534ce1d
  - `alertRule` -> `ngmodels.AlertRule`
- Alertmanager Configuration: 9f5b6b2ed2efb9f6bfeb41de7c1172938534ce1d
  - `PostableUserConfig`, `PostableApiAlertingConfig`, `PostableApiReceiver`, `Route`, `ObjectMatchers`, `PostableGrafanaReceiver` to their `apimodels` equivalents of the same name. 
- Secure settings encryption: 149ec38d6871433c9d2d4157f05f9562d43d9cf7
  -  `util.Encrypt` ->  `secrets.Service.Encrypt`
- Legacy notification channels: 42dcb8009510eaeebb10a2a8995ae250e09551f9
  - `notificationChannel` -> `legacymodels.AlertNotification`
- Dashboards:
  - `dashboard` -> `dashboards.Dashboard` d0230235e0e2cd3ac30fe3db5acfd53669b93170
  - Raw SQL -> services a41f32dd89b939c0b4db1825c1510510a9c08f47
-  Datasources: 53b9e478cdfdef680dc8ca889bb3067263c0216d
  - `dsUIDLookup` -> `datasources.CacheService`

Some gaps in the testing suite are filled:

- Migration of alert rules: verifying that the actual data model (queries, conditions) are correct 9a7cfa959171e60d97061d12f7ecbb5317059485
- Secure settings migration: verifying that secure fields remain encrypted for all available notifiers and certain fields migrate from plain text to encrypted secure settings correctly e7d399368816cf06ec88712173df655340f00d3a

Replacing the checks for custom dashboard ACLs will be done in a separate targeted PR as it adds functionality instead of moving it around.

JacobsonMT committed:
Move necessary ualert code to ngalert

JacobsonMT committed:
Refactor migration to not need migrator.Migrator

JacobsonMT committed:
Hook up new migration service to startup

Run before ngalert startup.
Use kvstore to check/save whether migration has been run.
Force run dashboard provisioning before migration.
Let dashboard provisioner provision legacy alerts even if UA is enabled.

JacobsonMT committed:
Remove condition from AddDashboardUIDPanelIDMigration, add to MigrationService

JacobsonMT committed:
Remove condition from RerunDashAlertMigration and give it a clearer name

JacobsonMT committed:
Remove condition from extractAlertmanagerConfigurationHistory

This always ran before the main migration, so it never did anything to new
legacy migrations.

JacobsonMT committed:
Remove condition from UpdateRuleGroupIndexMigration

This fixes RuleGroupIndex for existing rules to be incremental. Not necessary
to add this logic to MigrationService since each alert gets its own group.

JacobsonMT committed:
Remove CreateDefaultFoldersForAlertingMigration and other unused ones

CreateDefaultFoldersForAlertingMigration no longer has a purpose as the legacy
migration will create the general alerting folder when necessary. Also, creating
a new folder via the rule creation page is simpler than when this migration was
created.

Removes some leftover unused migration code.

JacobsonMT committed:
Fix migration tests to work with new design

JacobsonMT committed:
Remove provisioning before migration for now to keep PR targeted

JacobsonMT committed:
Move MigrationService to its own BackgroundService and revert changes to ngalert

MigrationService will no longer be guaranteed to run before ngalert starts, but
it should only cause a slight delay as the scheduler catches up after migration.

JacobsonMT committed:
Add migration to copy legacy migration status to kvstore

Also, replaces kvstore.AllOrganizations with 0 as kvstore.AllOrganizations has
special meaning when using kvstore.Get.

JacobsonMT committed:
Fix migration_test

JacobsonMT committed:
Replace vendored alertRule and PostableUserConfig with ngalert versions

yuri-tceretian commented on pkg/services/sqlstore/migrations/migrations.go:
```diff
@@ -96,6 +94,8 @@ func (*OSSMigrations) AddMigration(mg *Migrator) {
 		}
 	}
+
+	ualert.MigrationServiceMigration(mg)
```
Should we leave a comment to always keep it at the bottom?

yuri-tceretian commented on pkg/services/ngalert/migration/service.go:
```diff
@@ -0,0 +1,200 @@
+
+		// Safeguard to prevent data loss when migrating from UA to LA
+		if !ms.cfg.ForceMigration {
+			panic("Grafana has already been migrated to Unified Alerting.\nAny alert rules created while using Unified Alerting will be deleted by rolling back.\n\nSet force_migration=true in your grafana.ini and restart Grafana to roll back and delete Unified Alerting configuration data.")
```
I think it can be a regular error. The service is running as part of the error Group anyway.

JacobsonMT replied:
Ah, true. Nice catch!

yuri-tceretian commented on pkg/services/ngalert/migration/service.go:
```diff
@@ -0,0 +1,200 @@
+
+		// Since both revert and this modifies kv_store, we use the same transaction in both writes to prevent Lock Wait Timeout.
+		withTransaction := context.WithValue(ctx, sqlstore.ContextSessionKey{}, dbSess)
+		err = ms.SetMigrated(withTransaction, false)
```
I think propagation of the context would be a better choice. The framework will reuse the transaction from the context

JacobsonMT replied:
Agreed, but `WithTransactionalDbSession` doesn't store the session in context afaik, hence why I added it manually. Possibly a bug?

JacobsonMT replied:
Ah nvm, should call `InTransaction` first and then next the `WithTransactionalDbSession` :+1:

yuri-tceretian commented on pkg/services/ngalert/migration/testing.go:
```diff
@@ -1,22 +1,19 @@
-
-			Logger: log.New("test"),
-		},
+		log: log.New("test"),
```
just fyi, we have `logtest.Fake`

JacobsonMT committed:
Add more unit tests for alert rules migration

JacobsonMT committed:
Improve testing for migration of channel secure settings

JacobsonMT committed:
Replace vendored notificationChannel with AlertNotification

JacobsonMT committed:
Encrypt migrated secureSettings with secrets.Service

JacobsonMT committed:
Replace vendored dashboard with dashboards.Dashboard

JacobsonMT committed:
Improve testing for creating new folders when migrating

JacobsonMT committed:
Replace vendored dashboard & sql with dashboards.Dashboard and services

Leaves custom ACL logic for followup as it needs to be completely rewritten for
rbac.

JacobsonMT committed:
Replace raw sql datasource lookup with datasources.CacheService

JacobsonMT committed:
Propagate session through context

JacobsonMT committed:
Add ServerLockService to support HA

JacobsonMT committed:
Fix migration_test

JacobsonMT committed:
Replace panic with error on revert without force_migration config

JacobsonMT committed:
Use logtest.Fake

JacobsonMT commented on pkg/services/ngalert/migration/alert_rule.go:
```diff
@@ -27,8 +28,10 @@ type alertRule struct {
-	NamespaceUID    string `xorm:"namespace_uid"`
+	UID             string  `xorm:"uid"`
+	NamespaceUID    string  `xorm:"namespace_uid"`
+	DashboardUID    *string `xorm:"dashboard_uid"`
```
Added because `AddDashboardUIDPanelIDMigration` used to be conditional on migration (runs and reverts with migration). Since we cna't rely on it anymore the migration service needs to populate the fields. See 2fdf1d06e25939ad9ef623f21fe37b96078420ed

JacobsonMT commented on pkg/services/sqlstore/migrations/ualert/ualert.go:
```diff
@@ -54,99 +27,16 @@ func (e MigrationError) Error() string {
-
+// FixEarlyMigration fixes UA configs created before 8.2 with org_id=0 and moves some files like __default__.tmpl.
+// The only use of this migration is when a user enabled ng-alerting before 8.2.
+func FixEarlyMigration(mg *migrator.Migrator) {
```
This used to be called `RerunDashAlertMigration` which was wholly confusing as it didn't actually rerun the migration. Renamed to better represent what it was doing, and removed the condition on migration.

The code is not added to the migration service as it's no longer relevant. See 60d35fbbcf32e0e5d0e73e9da5347ed4d0e6c610

JacobsonMT commented on pkg/services/sqlstore/migrations/ualert/tables.go:
```diff
@@ -492,9 +492,6 @@ func addAlertImageMigrations(mg *migrator.Migrator) {
 }
 
 func extractAlertmanagerConfigurationHistoryMigration(mg *migrator.Migrator) {
-	if !mg.Cfg.UnifiedAlerting.IsEnabled() {
```
This always ran before the migration, so it never did anything to new
legacy migrations anyways.

JacobsonMT commented on pkg/services/sqlstore/migrations/ualert/rule_group_index_migration.go:
```diff
@@ -0,0 +1,209 @@
+
+// UpdateRuleGroupIndexMigration updates a new field rule_group_index for alert rules that belong to a group with more than 1 alert.
+func UpdateRuleGroupIndexMigration(mg *migrator.Migrator) {
+	mg.AddMigration("update group index for alert rules", &updateRulesOrderInGroup{})
```
Moved as-is from ualert along with the vendored structs it needed from elsewhere in the migration model.

The only change is here is removing the [condition to only add if UA is enabled.](https://github.com/grafana/grafana/blob/58f6648505d1fbc05d6fc9ee0b37720a9f3c3f5d/pkg/services/sqlstore/migrations/ualert/ualert.go#L807-L809)

JacobsonMT commented on pkg/services/sqlstore/migrations/ualert/ualert.go:
```diff
@@ -735,191 +226,3 @@ func (u *upgradeNgAlerting) updateAlertmanagerFiles(orgId int64, migrator *migra
-}
-
-// CreateDefaultFoldersForAlertingMigration creates a folder dedicated for alerting if no folders exist
-func CreateDefaultFoldersForAlertingMigration(mg *migrator.Migrator) {
```
This migration was removed entirely as it's no longer necessary.

- The migration service will create the general alerting folder when ran.
- Creating a new first folder via the rule creation page is simpler than when this migration was created.

JacobsonMT commented on pkg/services/ngalert/tests/fakes/kvstore.go:
```diff
@@ -0,0 +1,102 @@
+package fakes
```
Moved as-is from notifier.testing so it can be used for migration tests.

JacobsonMT commented on pkg/services/sqlstore/migrations/ualert/migration_service_migration.go:
```diff
@@ -0,0 +1,69 @@
+var migratedKey = "migrated"
+
+// MigrationServiceMigration moves the legacy alert migration status from the migration log to kvstore.
+func MigrationServiceMigration(mg *migrator.Migrator) {
```
This is necessary so that we retain the current migration state between when sqlstore was in charge of UA migration and now. See 79781aefcc995be95c2dff8b02099aff39abf8af

JacobsonMT reviewed (commented):
Some review notes

JacobsonMT commented on pkg/services/ngalert/migration/permissions.go:
```diff
@@ -0,0 +1,300 @@
+package migration
```
This is moved nearly 100% as-is from sqstore, not sure why it's not detected as a move. The only changes are to use the session through the context instead of through `folderHelper`. Browser diff: https://www.diffchecker.com/ldxQCY9E/

JacobsonMT commented on pkg/services/sqlstore/migrations/external_alertmanagers.go:
```diff
@@ -124,3 +126,22 @@ func generateNewDatasourceUid(sess *xorm.Session, orgId int64) (string, error) {
+type secureJsonData map[string][]byte
+
+// getEncryptedJsonData returns map where all keys are encrypted.
+func getEncryptedJsonData(sjd map[string]string, log log.Logger) secureJsonData {
```
Leftover from `securejsondata.go`

JacobsonMT reviewed (commented):
Review notes part 2

JacobsonMT committed:
Skip error linting for service-level ForceMigrationError

JacobsonMT committed:
Linting

JacobsonMT committed:
Revert accidental change

JacobsonMT committed:
Assert accesscontrol.BackgroundUser back to user.SignedInUser

JacobsonMT commented:
Updated main and rebased

JacobsonMT committed:
Fix regression: unknown datasource id errors instead of uses empty uid

JacobsonMT committed:
Noop refactor - combine most external store access into migration/store

JacobsonMT committed:
Fix merge conflict with ed7d29f2b98e0e1221327a55a4096499ef78d3a2

yuri-tceretian reviewed (approved):
LGTM. I reviewed it as part of #74503

JacobsonMT committed:
Remove vendored dashAlert

JacobsonMT committed:
Refactor migration to be per org

This is a needed for upcoming work, so we do it separately for clarity.

JacobsonMT committed:
Keep track of new folders using kvstore instead of dashboard.created_by

Setting dashboard.created_by to -8 as a way to track which folders were
created by the migration is no longer possible. Instead, we store the
newly created folders uids in the kvstore and use it on revert.

JacobsonMT committed:
Fix linting

JacobsonMT committed:
Fix possible deadlock in migration_service_migration

If migrationLock feature flag was enabled and the max db
connections was 2 or less, a deadlock would occur. This is
because the GetMigrationLog() method spawns a new session
and when migrationLock is used, the migration is executed
in a nested transaction. Together, these deplete the
connection pool.

JacobsonMT commented:
/deploy-to-hg

JacobsonMT commented:
/deploy-to-hg

yuri-tceretian reviewed (approved):
LGTM
